File: csharp/CsvReading.cs
using System.Globalization;

namespace AllocationChallenge;

public static class CsvReading
{
    public static List<ConsumptionVolume> LoadConsumption(string filePath)
    {
        var volumes = new List<ConsumptionVolume>();

        foreach (var record in ReadCsvRecords(filePath))
        {
            var volume = new ConsumptionVolume(
                Consumer: record["consumer"],
                PeriodStart: ParseTimestamp(record["timestamp"]),
                Duration: Duration.Hour,
                QuantityKwh: decimal.Parse(record["quantity"], CultureInfo.InvariantCulture)
            );
            volumes.Add(volume);
        }

        return volumes;
    }

    public static List<ProductionVolume> LoadProduction(string filePath)
    {
        var volumes = new List<ProductionVolume>();

        foreach (var record in ReadCsvRecords(filePath))
        {
            var volume = new ProductionVolume(
                Producer: record["producer"],
                PeriodStart: ParseTimestamp(record["timestamp"]),
                Duration: Duration.Hour,
                QuantityKwh: decimal.Parse(record["quantity"], CultureInfo.InvariantCulture)
            );
            volumes.Add(volume);
        }

        return volumes;
    }

    public static List<AllocationCsvRow> LoadAllocations(string filePath)
    {
        var allocations = new List<AllocationCsvRow>();

        foreach (var record in ReadCsvRecords(filePath))
        {
            var allocation = new AllocationCsvRow(
                Producer: record["producer"],
                Consumer: record["consumer"],
                PeriodStart: ParseTimestamp(record["timestamp"]),
                Duration: Duration.Hour,
                QuantityKwh: decimal.Parse(record["quantity"], CultureInfo.InvariantCulture)
            );
            allocations.Add(allocation);
        }

        return allocations;
    }

    private static DateTime ParseTimestamp(string timestamp)
    {
        return DateTime.Parse(timestamp, CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal);
    }

    private static IEnumerable<Dictionary<string, string>> ReadCsvRecords(string filePath)
    {
        using var reader = new StreamReader(filePath, new System.Text.UTF8Encoding(encoderShouldEmitUTF8Identifier: true));

        var headerLine = reader.ReadLine();
        if (headerLine == null) yield break;

        // Remove BOM if present (StreamReader may not strip it in all cases)
        headerLine = headerLine.TrimStart('\uFEFF');

        var headers = headerLine.Split(',');

        while (reader.ReadLine() is { } line)
        {
            var values = line.Split(',');
            var record = new Dictionary<string, string>();

            for (int i = 0; i < headers.Length && i < values.Length; i++)
            {
                record[headers[i]] = values[i];
            }

            yield return record;
        }
    }
}

File: java/src/main/java/com/granularenergy/allocation/csv/CsvReading.java
package com.granularenergy.allocation.csv;

import com.granularenergy.allocation.models.ConsumptionVolume;
import com.granularenergy.allocation.models.Duration;
import com.granularenergy.allocation.models.ProductionVolume;
import com.opencsv.CSVReader;
import com.opencsv.exceptions.CsvValidationException;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CsvReading {

    private static final DateTimeFormatter TIMESTAMP_FORMATTER =
        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ssXXX");

    public static LocalDateTime parseTimestamp(String timestampStr) {
        OffsetDateTime odt = OffsetDateTime.parse(timestampStr, TIMESTAMP_FORMATTER);
        return odt.toLocalDateTime();
    }

    public static List<ConsumptionVolume> loadConsumption(Path filePath) throws IOException, CsvValidationException {
        List<ConsumptionVolume> consumptionVolumes = new ArrayList<>();

        try (CSVReader reader = createCsvReader(filePath)) {
            String[] headers = reader.readNext();
            Map<String, Integer> headerMap = createHeaderMap(headers);

            String[] row;
            while ((row = reader.readNext()) != null) {
                ConsumptionVolume volume = new ConsumptionVolume(
                    row[headerMap.get("consumer")],
                    parseTimestamp(row[headerMap.get("timestamp")]),
                    Duration.HOUR,
                    new BigDecimal(row[headerMap.get("quantity")])
                );
                consumptionVolumes.add(volume);
            }
        }

        return consumptionVolumes;
    }

    public static List<ProductionVolume> loadProduction(Path filePath) throws IOException, CsvValidationException {
        List<ProductionVolume> productionVolumes = new ArrayList<>();

        try (CSVReader reader = createCsvReader(filePath)) {
            String[] headers = reader.readNext();
            Map<String, Integer> headerMap = createHeaderMap(headers);

            String[] row;
            while ((row = reader.readNext()) != null) {
                ProductionVolume volume = new ProductionVolume(
                    row[headerMap.get("producer")],
                    parseTimestamp(row[headerMap.get("timestamp")]),
                    Duration.HOUR,
                    new BigDecimal(row[headerMap.get("quantity")])
                );
                productionVolumes.add(volume);
            }
        }

        return productionVolumes;
    }

    public static void loadAllocations(Path filePath) throws IOException, CsvValidationException {
        try (CSVReader reader = createCsvReader(filePath)) {
            String[] headers = reader.readNext();
            Map<String, Integer> headerMap = createHeaderMap(headers);

            String[] row;
            while ((row = reader.readNext()) != null) {
                String producer = row[headerMap.get("producer")];
                String consumer = row[headerMap.get("consumer")];
                LocalDateTime periodStart = parseTimestamp(row[headerMap.get("timestamp")]);
                Duration duration = Duration.HOUR;
                BigDecimal quantityKwh = new BigDecimal(row[headerMap.get("quantity")]);

                // TODO: return allocations
            }
        }
    }

    public static Path getRepoRoot() {
        return getRepoRoot(Path.of("").toAbsolutePath());
    }

    public static Path getRepoRoot(Path currentDir) {
        if (currentDir.resolve(".git").toFile().isDirectory()) {
            return currentDir;
        }
        Path parentDir = currentDir.getParent();
        if (parentDir == null || parentDir.equals(currentDir)) {
            throw new RuntimeException("Repository root not found");
        }
        return getRepoRoot(parentDir);
    }

    private static CSVReader createCsvReader(Path filePath) throws IOException {
        FileInputStream fis = new FileInputStream(filePath.toFile());
        InputStreamReader isr = new InputStreamReader(fis, StandardCharsets.UTF_8);
        return new CSVReader(isr);
    }

    private static Map<String, Integer> createHeaderMap(String[] headers) {
        Map<String, Integer> headerMap = new HashMap<>();
        for (int i = 0; i < headers.length; i++) {
            // Strip BOM if present on first header
            String header = headers[i];
            if (i == 0 && header.startsWith("\uFEFF")) {
                header = header.substring(1);
            }
            headerMap.put(header, i);
        }
        return headerMap;
    }
}
